### PR TITLE
Use native lib in unit tests

### PR DESCRIPTION
**Pull request by mdrlzy**

### CODE REVIEW

File: arklib/Cargo.toml [hunk 1]
```diff
@@ -15,8 +15,8 @@ url = { version = "2.2.2", features = ["serde"] }
 openssl = { version = "0.10.42", features = ["vendored"] }
 image = "0.24.2"
 log = "0.4.14"
-
-[target.'cfg(target_os="android")'.dependencies]
 jni = { version = "0.19.0", default-features = false }
 android_logger = "0.11.0"
+
+[target.'cfg(target_os="android")'.dependencies]
 ndk = "0.6.0"
```
Comment (kirillt) [inline arklib/Cargo.toml:22]: Am I right that NDK is useless on desktop?

File: scripts/buildRustLibForHost.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+fi
+
+# Mac
+if [ -e "target/$host_arch/debug/libarklib.dylib" ]; then
+    lib_file="target/$host_arch/debug/libarklib.dylib"
+fi
+
+# .dll for Windows
+
+cp $lib_file ../
```
Comment (kirillt) [inline scripts/buildRustLibForHost.sh:19]: Should be covered in README.

File: scripts/buildRustLibForHost.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
Comment (kirillt) [inline scripts/buildRustLibForHost.sh:1]: :+1:

File: lib/src/test/java/space/taran/arklib/ResourceIdTest.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    @Test
+    fun resourceIdTest() {
+        System.loadLibrary("arklib")
+        val file = createTempFile()
+        file.writeText("hello")
+        assertEquals(
+            computeId(file.fileSize(), file).toString(),
+            "5-907060870"
+        )
+    }
```
Comment (kirillt) [inline lib/src/test/java/space/taran/arklib/ResourceIdTest.kt:20]: Let's move `System.loadLibrary` into `@Before` function and create one more small test case with computing id directly from bytes.

File: arklib/Cargo.toml (hunk 1 above)
Comment (mdrlzy) [reply arklib/Cargo.toml:22]: Yes, moreover there will be an error
` compile_error!("android-ndk-sys only supports compiling for Android");`

File: lib/src/test/java/space/taran/arklib/ResourceIdTest.kt (hunk 4 above)
Comment (mdrlzy) [reply lib/src/test/java/space/taran/arklib/ResourceIdTest.kt]: We have binding only for files

**Comment by mdrlzy** [comment]
Problems:

- It's not clear how to test `BaseStorage.persist()` because under the hood it uses flow and debounce
https://slack-chats.kotlinlang.org/t/489420/does-anyone-know-how-to-unit-test-code-which-uses-flow-debou

- It's not clear how to use prepared files (PDF, etc.)

**Comment by kirillt** [comment]
@mdrlzy am I right, that:
1. Previews rendering (`Bitmap` usage) and video metadata extraction (decoders) require Android SDK — possible to test only using instrumented tests?
2. Same for `BaseStorage.persist()`?

Regarding prepared PDF and other files — probably tiny samples are OK to keep in the repo.

Probably we can create Git `submodule`, put all heavy stuff in there and use those samples only from instrumented tests as separate pipeline stage? The workflow would call something like `git submodule update --init --recursive` before running these heavier tests. @oluiscabral @antlko @gwendalF Guys, any idea regarding this?

File: lib/src/test/java/space/taran/arklib/lib/ResourceIdTest.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+class ResourceIdTest {
+    @Before
+    fun before() {
+        System.loadLibrary("arklib")
+    }
+
+    @Test
+    fun resourceIdTest() {
+        val file = createTempFile()
```
Comment (oluiscabral) [inline lib/src/test/java/space/taran/arklib/lib/ResourceIdTest.kt:19]: Although this file is temporary, it's a good practice to delete it at the end of the test.

File: lib/src/test/java/space/taran/arklib/user/tags/RootTagStorageTest.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    val root = TestFiles.root1
+    lateinit var index: ResourceIndex
+    lateinit var tagStorage: TagStorage
+
+    @Before
+    fun before() = runTest {
+        System.loadLibrary("arklib")
+        mockLog()
+        TestFiles.randomFile(root)
```
Comment (oluiscabral) [inline lib/src/test/java/space/taran/arklib/user/tags/RootTagStorageTest.kt:26]: Where is this random file being deleted?

**Review by oluiscabral — CHANGES_REQUESTED** [review]

**Comment by mdrlzy** [comment]
> @mdrlzy am I right, that:
> 
> 1. Previews rendering (`Bitmap` usage) and video metadata extraction (decoders) require Android SDK — possible to test only using instrumented tests?
> 2. Same for `BaseStorage.persist()`?

@kirillt 
1. Yes, this is part of Android SDK
2. The problem is that flushing to disk has a delay after request. I have not yet found a way to "wait" for this delay

File: lib/src/test/java/space/taran/arklib/lib/ResourceIdTest.kt (hunk 5 above)
Comment (mdrlzy) [reply lib/src/test/java/space/taran/arklib/lib/ResourceIdTest.kt:19]: Sure

File: lib/src/test/java/space/taran/arklib/user/tags/RootTagStorageTest.kt (hunk 6 above)
Comment (mdrlzy) [reply lib/src/test/java/space/taran/arklib/user/tags/RootTagStorageTest.kt:26]: Here a temporary file is also created, but I will add its removal

**Comment by kirillt** [comment]
@mdrlzy
1. Eventually, we'll port previews generation and video processing to `arklib`. Yes, we can't utilize unit tests to compare the two implementations, but it's not a critical issue for well-separated functionality.
2. Not sure, but there was this link to SO in the last post of that topic on Kotlin forum: https://stackoverflow.com/questions/61224047/unit-testing-coroutines-runblockingtest-this-job-has-not-completed-yet/67897542#67897542

File: lib/src/test/java/space/taran/arklib/utils/TestFiles.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    // 100% unique file name and content
+    private val generatedFileContent = mutableSetOf<String>()
+    fun randomFile(parent: Path, extension: String = ""): Path {
+        var uuid: String
+        do {
+            uuid = UUID.randomUUID().toString()
+        } while (generatedFileContent.contains(uuid))
+
+        generatedFileContent.add(uuid)
+        val file = parent.resolve("$uuid.$extension")
```
Comment (kirillt) [inline lib/src/test/java/space/taran/arklib/utils/TestFiles.kt:28]: `extension` can be empty

File: lib/src/test/resources/space/taran/arklib/lib/test.pdf
Comment (kirillt) [inline lib/src/test/resources/space/taran/arklib/lib/test.pdf:1]: Maybe we should already start hiding big files by putting them into `git submodule`.
Running tests would require something like `git submodule update --init --recursive`.

**Pull request merged**